src/services/codefixes/convertTypedefToType.ts

```diff
@@ -0,0 +1,204 @@
+    typeExpression: JSDocTypeExpression
+): TypeAliasDeclaration | undefined {
+    const typeReference = getSynthesizedDeepClone(typeExpression.type);
+    if (!typeReference) return;
+    const declaration = factory.createTypeAliasDeclaration(
+        [],
+        factory.createIdentifier(typeName),
+        [],
+        typeReference
+    );
+    return declaration;
+}
+
+function createSignatureFromTypeLiteral(typeLiteral: JSDocTypeLiteral): PropertySignature[] | undefined {
+    const propertyTags = typeLiteral.jsDocPropertyTags;
+    if (!some(propertyTags)) return;
+
+    const getSignature = (tag: JSDocPropertyTag) => {
+        const name = getPropertyName(tag);
+        const type = tag.typeExpression?.type;
+        const isOptional = tag.isBracketed;
+        let typeReference;
+
+        // Recursively handle nested object type
+        if (type && type.kind === SyntaxKind.JSDocTypeLiteral) {
+            const signatures = createSignatureFromTypeLiteral(type as JSDocTypeLiteral);
+            typeReference = factory.createTypeLiteralNode(signatures);
+        }
+        // Leaf node, where type.kind === SyntaxKind.JSDocTypeExpression
+        else if (type) {
+            typeReference = getSynthesizedDeepClone(type);
+        }
+
+        if (typeReference && name) {
+            const questionToken = isOptional ? factory.createToken(SyntaxKind.QuestionToken) : undefined;
+            const prop = factory.createPropertySignature(
+                [],
+                name,
+                questionToken,
+                typeReference
+            );
+
+            return prop;
+        }
+    };
+
+    const props = mapDefined(propertyTags as JSDocPropertyTag[], getSignature);
+    return props;
+}
+
+function getPropertyName(tag: JSDocPropertyTag): string | undefined {
+    const { name } = tag;
+    if (!name) return;
+
+    let propertyName;
+    // for "@property {string} parent.child" or "@property {string} parent.child.grandchild" in nested object type
+    // We'll get "child" in the first example or "grandchild" in the second example as the prop name
+    if (name.kind === SyntaxKind.QualifiedName) {
+        propertyName = name.right.text;
+    }
+    else {
+        propertyName = tag.name.getText();
+    }
+    return propertyName;
+}
+
+
+/** @internal */
+export function containsJSDocTypedef(node: Node): node is HasJSDoc {
+    return hasJSDocNodes(node) && some(node.jsDoc, node => some(node.tags, tag => isJSDocTypedefTag(tag)));
+}
+
+/** @internal */
+export function getJSDocTypedefNode(node: HasJSDoc): JSDocTypedefTag {
+    const jsDocNodes = node.jsDoc || [];
+
+    return flatMap(jsDocNodes, (node) => {
+        const tags = node.tags || [];
+        return tags.filter((tag) => isJSDocTypedefTag(tag));
+    })[0] as unknown as JSDocTypedefTag;
+}
+
+/** @internal */
+export function containsTypeDefTag(node: Node): node is JSDocTypedefTag {
+    return isInJSDoc(node) && isJSDocTypedefTag(node);
+}
```

Comment:
This function seems to be named incorrectly, and isn’t it redundant with just `isJSDocTypedefTag`? I didn’t think it was possible for something to pass `isJSDocTypedefTag` without implying `isInJSDoc`.

Reply:
changed